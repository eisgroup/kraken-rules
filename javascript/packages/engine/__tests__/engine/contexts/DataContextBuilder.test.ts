/* eslint-disable @typescript-eslint/no-non-null-assertion */
/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { DataContextBuilder } from '../../../src/engine/contexts/data/DataContextBuilder'
import { mock } from '../../mock'
import { DataContext } from '../../../src/engine/contexts/data/DataContext'

describe('DataContextBuilder', () => {
    const builder = new DataContextBuilder(mock.modelTree, mock.spi.instance)
    const dataObject = Object.freeze(mock.data.empty())
    const { Insured, Party, AddressInfo } = mock.modelTreeJson.contexts
    it('should create instance', () => {
        expect(builder).not.toBeNull()
    })
    it('should check created instance', () => {
        const dataContext: DataContext = builder.buildFromRoot(dataObject)
        expect(dataContext.dataObject).toMatchObject(dataObject)
        expect(dataContext.contextName).toBe(dataObject['cd'])
        expect(dataContext.contextId).toBeDefined()
    })
    it('should build from extracted array first element', () => {
        const parent: DataContext = builder.buildFromRoot(dataObject)
        const party = builder.buildFromExtractedObject(dataObject.parties![0], Party.name, parent, 0)
        expect(party).toMatchSnapshot()
    })
    it('should build from extracted object', () => {
        const parent: DataContext = builder.buildFromRoot(dataObject)
        const human0DataContext = builder.buildFromExtractedObject(dataObject.insured!, Insured.name, parent, 0)
        const dataContext = builder.buildFromExtractedObject(
            dataObject.insured!.addressInfo!,
            AddressInfo.name,
            human0DataContext,
        )
        expect(dataContext).toMatchSnapshot()
    })
    it('should throw on undefined data object', () => {
        expect(() => builder.buildFromRoot(undefined!)).toThrowError()
    })
    it('should throw on empty data object', () => {
        expect(() => builder.buildFromRoot({})).toThrowError()
    })
    it('should throw on empty info fields', () => {
        const parent: DataContext = builder.buildFromRoot(dataObject)
        expect(() => builder.buildFromExtractedObject({}, 'Policy', parent, 2)).toThrowError()
    })
})
