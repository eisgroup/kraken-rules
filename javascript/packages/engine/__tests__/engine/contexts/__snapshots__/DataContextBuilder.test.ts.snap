// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataContextBuilder should build from extracted array first element 1`] = `
DataContext {
  "contextDefinition": Object {
    "children": Object {
      "DriverInfo": Object {
        "cardinality": "SINGLE",
        "navigationExpression": Object {
          "expressionString": "driverInfo",
          "expressionType": "PATH",
          "expressionVariables": Array [],
          "originalExpressionString": "driverInfo",
        },
        "targetName": "DriverInfo",
      },
      "PartyRole": Object {
        "cardinality": "MULTIPLE",
        "navigationExpression": Object {
          "expressionString": "roles",
          "expressionType": "PATH",
          "expressionVariables": Array [],
          "originalExpressionString": "roles",
        },
        "targetName": "PartyRole",
      },
      "PersonInfo": Object {
        "cardinality": "SINGLE",
        "navigationExpression": Object {
          "expressionString": "personInfo",
          "expressionType": "PATH",
          "expressionVariables": Array [],
          "originalExpressionString": "personInfo",
        },
        "targetName": "PersonInfo",
      },
    },
    "fields": Object {
      "relationToPrimaryInsured": Object {
        "cardinality": "SINGLE",
        "fieldPath": "relationToPrimaryInsured",
        "fieldType": "STRING",
        "name": "relationToPrimaryInsured",
      },
      "roles": Object {
        "cardinality": "MULTIPLE",
        "fieldPath": "roles",
        "fieldType": "PartyRole",
        "name": "roles",
      },
    },
    "inheritedContexts": Array [],
    "name": "Party",
    "system": false,
  },
  "contextId": "3",
  "contextName": "Party",
  "contextPath": "path.to.3",
  "dataContextReferences": Object {
    "Party": Object {
      "cardinality": "SINGLE",
      "dataContexts": Array [
        [Circular],
      ],
      "name": "Party",
    },
  },
  "dataObject": Object {
    "cd": "Party",
    "id": "3",
    "personInfo": Object {
      "cd": "PersonInfo",
      "id": "4",
    },
    "roles": Array [
      Object {
        "cd": "PartyRole",
        "id": "5",
      },
    ],
  },
  "description": "Party:path.to.3:3",
  "id": "Party:3",
  "info": Object {
    "getContextInstanceId": [Function],
    "getContextName": [Function],
  },
  "objectReferences": Object {
    "Party": Object {
      "cd": "Party",
      "id": "3",
      "personInfo": Object {
        "cd": "PersonInfo",
        "id": "4",
      },
      "roles": Array [
        Object {
          "cd": "PartyRole",
          "id": "5",
        },
      ],
    },
  },
  "parent": DataContext {
    "contextDefinition": Object {
      "children": Object {
        "CarCoverage": Object {
          "cardinality": "SINGLE",
          "navigationExpression": Object {
            "expressionString": "coverage",
            "expressionType": "PATH",
            "expressionVariables": Array [],
            "originalExpressionString": "coverage",
          },
          "targetName": "CarCoverage",
        },
        "CreditCardInfo": Object {
          "cardinality": "SINGLE",
          "navigationExpression": Object {
            "expressionString": "this._flatMap(__dataObject__.billingInfo,_x_=>_x_.creditCardInfo)",
            "expressionType": "COMPLEX",
            "expressionVariables": Array [],
            "originalExpressionString": "billingInfo.creditCardInfo",
          },
          "targetName": "CreditCardInfo",
        },
        "Insured": Object {
          "cardinality": "SINGLE",
          "navigationExpression": Object {
            "expressionString": "insured",
            "expressionType": "PATH",
            "expressionVariables": Array [],
            "originalExpressionString": "insured",
          },
          "targetName": "Insured",
        },
        "Party": Object {
          "cardinality": "MULTIPLE",
          "navigationExpression": Object {
            "expressionString": "parties",
            "expressionType": "PATH",
            "expressionVariables": Array [],
            "originalExpressionString": "parties",
          },
          "targetName": "Party",
        },
        "Referer": Object {
          "cardinality": "SINGLE",
          "navigationExpression": Object {
            "expressionString": "referer",
            "expressionType": "PATH",
            "expressionVariables": Array [],
            "originalExpressionString": "referer",
          },
          "targetName": "Referer",
        },
        "SecondaryInsured": Object {
          "cardinality": "MULTIPLE",
          "navigationExpression": Object {
            "expressionString": "[__dataObject__.oneInsured,__dataObject__.multipleInsureds,[__dataObject__.multiInsureds1,__dataObject__.multiInsureds2]]",
            "expressionType": "COMPLEX",
            "expressionVariables": Array [],
            "originalExpressionString": "{ oneInsured, multipleInsureds, { multiInsureds1, multiInsureds2 }}",
          },
          "targetName": "SecondaryInsured",
        },
        "TransactionDetails": Object {
          "cardinality": "SINGLE",
          "navigationExpression": Object {
            "expressionString": "transactionDetails",
            "expressionType": "PATH",
            "expressionVariables": Array [],
            "originalExpressionString": "transactionDetails",
          },
          "targetName": "TransactionDetails",
        },
        "Vehicle": Object {
          "cardinality": "MULTIPLE",
          "navigationExpression": Object {
            "expressionString": "riskItems",
            "expressionType": "PATH",
            "expressionVariables": Array [],
            "originalExpressionString": "riskItems",
          },
          "targetName": "Vehicle",
        },
      },
      "fields": Object {
        "accountName": Object {
          "cardinality": "SINGLE",
          "fieldPath": "billingInfo.accountName",
          "fieldType": "STRING",
          "name": "accountName",
        },
        "backupSystemDate": Object {
          "cardinality": "SINGLE",
          "fieldPath": "backupSystemDate",
          "fieldType": "SysDate",
          "name": "backupSystemDate",
        },
        "changePremium": Object {
          "cardinality": "SINGLE",
          "fieldPath": "transactionDetails.changePremium",
          "fieldType": "DECIMAL",
          "name": "changePremium",
        },
        "contractTermTypeCd": Object {
          "cardinality": "SINGLE",
          "fieldPath": "termDetails.contractTermTypeCd",
          "fieldType": "STRING",
          "name": "contractTermTypeCd",
        },
        "createdBy": Object {
          "cardinality": "SINGLE",
          "fieldPath": "accessTrackInfo.createdBy",
          "fieldType": "STRING",
          "name": "createdBy",
        },
        "createdFromPolicyRev": Object {
          "cardinality": "SINGLE",
          "fieldPath": "createdFromPolicyRev",
          "fieldType": "INTEGER",
          "name": "createdFromPolicyRev",
        },
        "createdOn": Object {
          "cardinality": "SINGLE",
          "fieldPath": "accessTrackInfo.createdOn",
          "fieldType": "DATE",
          "name": "createdOn",
        },
        "currentQuoteInd": Object {
          "cardinality": "SINGLE",
          "fieldPath": "policyDetail.currentQuoteInd",
          "fieldType": "BOOLEAN",
          "name": "currentQuoteInd",
        },
        "effectiveDate": Object {
          "cardinality": "SINGLE",
          "fieldPath": "termDetails.termEffectiveDate",
          "fieldType": "DATE",
          "name": "effectiveDate",
        },
        "expirationDate": Object {
          "cardinality": "SINGLE",
          "fieldPath": "termDetails.termExpirationDate",
          "fieldType": "DATE",
          "name": "expirationDate",
        },
        "numberOfInsureds": Object {
          "cardinality": "MULTIPLE",
          "fieldPath": "numberOfInsureds",
          "fieldType": "INTEGER",
          "name": "numberOfInsureds",
        },
        "oosProcessingStage": Object {
          "cardinality": "SINGLE",
          "fieldPath": "policyDetail.oosProcessingStage",
          "fieldType": "STRING",
          "name": "oosProcessingStage",
        },
        "parties": Object {
          "cardinality": "MULTIPLE",
          "fieldPath": "parties",
          "fieldType": "Party",
          "name": "parties",
        },
        "policies": Object {
          "cardinality": "MULTIPLE",
          "fieldPath": "policies",
          "fieldType": "STRING",
          "name": "policies",
        },
        "policyCurrency": Object {
          "cardinality": "SINGLE",
          "fieldPath": "policyCurrency",
          "fieldType": "STRING",
          "name": "policyCurrency",
        },
        "policyNumber": Object {
          "cardinality": "SINGLE",
          "fieldPath": "policyNumber",
          "fieldType": "STRING",
          "name": "policyNumber",
        },
        "policyValue": Object {
          "cardinality": "SINGLE",
          "fieldPath": "policyValue",
          "fieldType": "MONEY",
          "name": "policyValue",
        },
        "refToCustomer": Object {
          "cardinality": "SINGLE",
          "fieldPath": "refToCustomer",
          "fieldType": "UNKNOWN",
          "name": "refToCustomer",
        },
        "referer": Object {
          "cardinality": "SINGLE",
          "fieldPath": "referer",
          "fieldType": "Referer",
          "name": "referer",
        },
        "riskItems": Object {
          "cardinality": "MULTIPLE",
          "fieldPath": "riskItems",
          "fieldType": "Vehicle",
          "name": "riskItems",
        },
        "state": Object {
          "cardinality": "SINGLE",
          "fieldPath": "state",
          "fieldType": "STRING",
          "name": "state",
        },
        "systemDate": Object {
          "cardinality": "SINGLE",
          "fieldPath": "systemDate",
          "fieldType": "SysDate",
          "name": "systemDate",
        },
        "termCd": Object {
          "cardinality": "SINGLE",
          "fieldPath": "termDetails.termCd",
          "fieldType": "STRING",
          "name": "termCd",
        },
        "termNo": Object {
          "cardinality": "SINGLE",
          "fieldPath": "termDetails.termNo",
          "fieldType": "INTEGER",
          "name": "termNo",
        },
        "totalLimit": Object {
          "cardinality": "SINGLE",
          "fieldPath": "transactionDetails.totalLimit",
          "fieldType": "DECIMAL",
          "name": "totalLimit",
        },
        "totalPremium": Object {
          "cardinality": "SINGLE",
          "fieldPath": "transactionDetails.totalPremium",
          "fieldType": "DECIMAL",
          "name": "totalPremium",
        },
        "txCreateDate": Object {
          "cardinality": "SINGLE",
          "fieldPath": "transactionDetails.txCreateDate",
          "fieldType": "DATE",
          "name": "txCreateDate",
        },
        "txEffectiveDate": Object {
          "cardinality": "SINGLE",
          "fieldPath": "transactionDetails.txEffectiveDate",
          "fieldType": "DATETIME",
          "name": "txEffectiveDate",
        },
        "txReason": Object {
          "cardinality": "SINGLE",
          "fieldPath": "transactionDetails.txReason",
          "fieldType": "STRING",
          "name": "txReason",
        },
        "txType": Object {
          "cardinality": "SINGLE",
          "fieldPath": "transactionDetails.txType",
          "fieldType": "STRING",
          "name": "txType",
        },
        "updatedBy": Object {
          "cardinality": "SINGLE",
          "fieldPath": "accessTrackInfo.updatedBy",
          "fieldType": "STRING",
          "name": "updatedBy",
        },
        "updatedOn": Object {
          "cardinality": "SINGLE",
          "fieldPath": "accessTrackInfo.updatedOn",
          "fieldType": "DATE",
          "name": "updatedOn",
        },
        "versionDescription": Object {
          "cardinality": "SINGLE",
          "fieldPath": "policyDetail.versionDescription",
          "fieldType": "STRING",
          "name": "versionDescription",
        },
      },
      "inheritedContexts": Array [],
      "name": "Policy",
      "system": false,
    },
    "contextId": "0",
    "contextName": "Policy",
    "contextPath": "path.to.0",
    "dataContextReferences": Object {
      "Policy": Object {
        "cardinality": "SINGLE",
        "dataContexts": Array [
          [Circular],
        ],
        "name": "Policy",
      },
    },
    "dataObject": Object {
      "accessTrackInfo": Object {
        "cd": "AccessTrackInfo",
        "id": "9",
      },
      "billingInfo": Object {
        "cd": "BillingInfo",
        "creditCardInfo": Object {
          "cd": "CreditCardInfo",
          "id": "2",
        },
        "id": "1",
      },
      "cd": "Policy",
      "id": "0",
      "insured": Object {
        "addressInfo": Object {
          "cd": "AddressInfo",
          "id": "iai1",
        },
        "cd": "Insured",
        "id": "insured-1-id",
      },
      "parties": Array [
        Object {
          "cd": "Party",
          "id": "3",
          "personInfo": Object {
            "cd": "PersonInfo",
            "id": "4",
          },
          "roles": Array [
            Object {
              "cd": "PartyRole",
              "id": "5",
            },
          ],
        },
      ],
      "policyDetail": Object {
        "cd": "PolicyDetail",
        "id": "11",
      },
      "riskItems": Array [
        Object {
          "addressInfo": Object {
            "cd": "AddressInfo",
            "id": "7",
          },
          "cd": "Vehicle",
          "id": "6",
        },
      ],
      "termDetails": Object {},
      "transactionDetails": Object {
        "cd": "TransactionDetails",
        "id": "8",
      },
    },
    "description": "Policy:path.to.0:0",
    "id": "Policy:0",
    "info": Object {
      "getContextInstanceId": [Function],
      "getContextName": [Function],
    },
    "objectReferences": Object {
      "Policy": Object {
        "accessTrackInfo": Object {
          "cd": "AccessTrackInfo",
          "id": "9",
        },
        "billingInfo": Object {
          "cd": "BillingInfo",
          "creditCardInfo": Object {
            "cd": "CreditCardInfo",
            "id": "2",
          },
          "id": "1",
        },
        "cd": "Policy",
        "id": "0",
        "insured": Object {
          "addressInfo": Object {
            "cd": "AddressInfo",
            "id": "iai1",
          },
          "cd": "Insured",
          "id": "insured-1-id",
        },
        "parties": Array [
          Object {
            "cd": "Party",
            "id": "3",
            "personInfo": Object {
              "cd": "PersonInfo",
              "id": "4",
            },
            "roles": Array [
              Object {
                "cd": "PartyRole",
                "id": "5",
              },
            ],
          },
        ],
        "policyDetail": Object {
          "cd": "PolicyDetail",
          "id": "11",
        },
        "riskItems": Array [
          Object {
            "addressInfo": Object {
              "cd": "AddressInfo",
              "id": "7",
            },
            "cd": "Vehicle",
            "id": "6",
          },
        ],
        "termDetails": Object {},
        "transactionDetails": Object {
          "cd": "TransactionDetails",
          "id": "8",
        },
      },
    },
    "parent": undefined,
  },
}
`;

exports[`DataContextBuilder should build from extracted object 1`] = `
DataContext {
  "contextDefinition": Object {
    "children": Object {
      "AddressLine1": Object {
        "cardinality": "SINGLE",
        "navigationExpression": Object {
          "expressionString": "addressLine1",
          "expressionType": "PATH",
          "expressionVariables": Array [],
          "originalExpressionString": "addressLine1",
        },
        "targetName": "AddressLine1",
      },
      "AddressLine2": Object {
        "cardinality": "SINGLE",
        "navigationExpression": Object {
          "expressionString": "addressLine2",
          "expressionType": "PATH",
          "expressionVariables": Array [],
          "originalExpressionString": "addressLine2",
        },
        "targetName": "AddressLine2",
      },
    },
    "fields": Object {
      "addressLines": Object {
        "cardinality": "MULTIPLE",
        "fieldPath": "addressLines",
        "fieldType": "AddressLine",
        "name": "addressLines",
      },
      "city": Object {
        "cardinality": "SINGLE",
        "fieldPath": "city",
        "fieldType": "STRING",
        "name": "city",
      },
      "countryCd": Object {
        "cardinality": "SINGLE",
        "fieldPath": "countryCd",
        "fieldType": "STRING",
        "name": "countryCd",
      },
      "doNotSolicit": Object {
        "cardinality": "SINGLE",
        "fieldPath": "doNotSolicit",
        "fieldType": "BOOLEAN",
        "name": "doNotSolicit",
      },
      "postalCode": Object {
        "cardinality": "SINGLE",
        "fieldPath": "postalCode",
        "fieldType": "STRING",
        "name": "postalCode",
      },
      "stateCd": Object {
        "cardinality": "SINGLE",
        "fieldPath": "stateCd",
        "fieldType": "STRING",
        "name": "stateCd",
      },
      "street": Object {
        "cardinality": "SINGLE",
        "fieldPath": "street",
        "fieldType": "STRING",
        "name": "street",
      },
    },
    "inheritedContexts": Array [],
    "name": "AddressInfo",
    "system": false,
  },
  "contextId": "iai1",
  "contextName": "AddressInfo",
  "contextPath": "path.to.iai1",
  "dataContextReferences": Object {
    "AddressInfo": Object {
      "cardinality": "SINGLE",
      "dataContexts": Array [
        [Circular],
      ],
      "name": "AddressInfo",
    },
  },
  "dataObject": Object {
    "cd": "AddressInfo",
    "id": "iai1",
  },
  "description": "AddressInfo:path.to.iai1:iai1",
  "id": "AddressInfo:iai1",
  "info": Object {
    "getContextInstanceId": [Function],
    "getContextName": [Function],
  },
  "objectReferences": Object {
    "AddressInfo": Object {
      "cd": "AddressInfo",
      "id": "iai1",
    },
  },
  "parent": DataContext {
    "contextDefinition": Object {
      "children": Object {
        "BillingAddress": Object {
          "cardinality": "SINGLE",
          "navigationExpression": Object {
            "expressionString": "addressInfo",
            "expressionType": "PATH",
            "expressionVariables": Array [],
            "originalExpressionString": "addressInfo",
          },
          "targetName": "BillingAddress",
        },
      },
      "fields": Object {
        "addressInfo": Object {
          "cardinality": "SINGLE",
          "fieldPath": "addressInfo",
          "fieldType": "BillingAddress",
          "name": "addressInfo",
        },
        "childrenAges": Object {
          "cardinality": "MULTIPLE",
          "fieldPath": "childrenAges",
          "fieldType": "INTEGER",
          "name": "childrenAges",
        },
        "haveChildren": Object {
          "cardinality": "SINGLE",
          "fieldPath": "haveChildren",
          "fieldType": "BOOLEAN",
          "name": "haveChildren",
        },
        "name": Object {
          "cardinality": "SINGLE",
          "fieldPath": "name",
          "fieldType": "STRING",
          "name": "name",
        },
      },
      "inheritedContexts": Array [],
      "name": "Insured",
      "system": false,
    },
    "contextId": "insured-1-id",
    "contextName": "Insured",
    "contextPath": "path.to.insured-1-id",
    "dataContextReferences": Object {
      "Insured": Object {
        "cardinality": "SINGLE",
        "dataContexts": Array [
          [Circular],
        ],
        "name": "Insured",
      },
    },
    "dataObject": Object {
      "addressInfo": Object {
        "cd": "AddressInfo",
        "id": "iai1",
      },
      "cd": "Insured",
      "id": "insured-1-id",
    },
    "description": "Insured:path.to.insured-1-id:insured-1-id",
    "id": "Insured:insured-1-id",
    "info": Object {
      "getContextInstanceId": [Function],
      "getContextName": [Function],
    },
    "objectReferences": Object {
      "Insured": Object {
        "addressInfo": Object {
          "cd": "AddressInfo",
          "id": "iai1",
        },
        "cd": "Insured",
        "id": "insured-1-id",
      },
    },
    "parent": DataContext {
      "contextDefinition": Object {
        "children": Object {
          "CarCoverage": Object {
            "cardinality": "SINGLE",
            "navigationExpression": Object {
              "expressionString": "coverage",
              "expressionType": "PATH",
              "expressionVariables": Array [],
              "originalExpressionString": "coverage",
            },
            "targetName": "CarCoverage",
          },
          "CreditCardInfo": Object {
            "cardinality": "SINGLE",
            "navigationExpression": Object {
              "expressionString": "this._flatMap(__dataObject__.billingInfo,_x_=>_x_.creditCardInfo)",
              "expressionType": "COMPLEX",
              "expressionVariables": Array [],
              "originalExpressionString": "billingInfo.creditCardInfo",
            },
            "targetName": "CreditCardInfo",
          },
          "Insured": Object {
            "cardinality": "SINGLE",
            "navigationExpression": Object {
              "expressionString": "insured",
              "expressionType": "PATH",
              "expressionVariables": Array [],
              "originalExpressionString": "insured",
            },
            "targetName": "Insured",
          },
          "Party": Object {
            "cardinality": "MULTIPLE",
            "navigationExpression": Object {
              "expressionString": "parties",
              "expressionType": "PATH",
              "expressionVariables": Array [],
              "originalExpressionString": "parties",
            },
            "targetName": "Party",
          },
          "Referer": Object {
            "cardinality": "SINGLE",
            "navigationExpression": Object {
              "expressionString": "referer",
              "expressionType": "PATH",
              "expressionVariables": Array [],
              "originalExpressionString": "referer",
            },
            "targetName": "Referer",
          },
          "SecondaryInsured": Object {
            "cardinality": "MULTIPLE",
            "navigationExpression": Object {
              "expressionString": "[__dataObject__.oneInsured,__dataObject__.multipleInsureds,[__dataObject__.multiInsureds1,__dataObject__.multiInsureds2]]",
              "expressionType": "COMPLEX",
              "expressionVariables": Array [],
              "originalExpressionString": "{ oneInsured, multipleInsureds, { multiInsureds1, multiInsureds2 }}",
            },
            "targetName": "SecondaryInsured",
          },
          "TransactionDetails": Object {
            "cardinality": "SINGLE",
            "navigationExpression": Object {
              "expressionString": "transactionDetails",
              "expressionType": "PATH",
              "expressionVariables": Array [],
              "originalExpressionString": "transactionDetails",
            },
            "targetName": "TransactionDetails",
          },
          "Vehicle": Object {
            "cardinality": "MULTIPLE",
            "navigationExpression": Object {
              "expressionString": "riskItems",
              "expressionType": "PATH",
              "expressionVariables": Array [],
              "originalExpressionString": "riskItems",
            },
            "targetName": "Vehicle",
          },
        },
        "fields": Object {
          "accountName": Object {
            "cardinality": "SINGLE",
            "fieldPath": "billingInfo.accountName",
            "fieldType": "STRING",
            "name": "accountName",
          },
          "backupSystemDate": Object {
            "cardinality": "SINGLE",
            "fieldPath": "backupSystemDate",
            "fieldType": "SysDate",
            "name": "backupSystemDate",
          },
          "changePremium": Object {
            "cardinality": "SINGLE",
            "fieldPath": "transactionDetails.changePremium",
            "fieldType": "DECIMAL",
            "name": "changePremium",
          },
          "contractTermTypeCd": Object {
            "cardinality": "SINGLE",
            "fieldPath": "termDetails.contractTermTypeCd",
            "fieldType": "STRING",
            "name": "contractTermTypeCd",
          },
          "createdBy": Object {
            "cardinality": "SINGLE",
            "fieldPath": "accessTrackInfo.createdBy",
            "fieldType": "STRING",
            "name": "createdBy",
          },
          "createdFromPolicyRev": Object {
            "cardinality": "SINGLE",
            "fieldPath": "createdFromPolicyRev",
            "fieldType": "INTEGER",
            "name": "createdFromPolicyRev",
          },
          "createdOn": Object {
            "cardinality": "SINGLE",
            "fieldPath": "accessTrackInfo.createdOn",
            "fieldType": "DATE",
            "name": "createdOn",
          },
          "currentQuoteInd": Object {
            "cardinality": "SINGLE",
            "fieldPath": "policyDetail.currentQuoteInd",
            "fieldType": "BOOLEAN",
            "name": "currentQuoteInd",
          },
          "effectiveDate": Object {
            "cardinality": "SINGLE",
            "fieldPath": "termDetails.termEffectiveDate",
            "fieldType": "DATE",
            "name": "effectiveDate",
          },
          "expirationDate": Object {
            "cardinality": "SINGLE",
            "fieldPath": "termDetails.termExpirationDate",
            "fieldType": "DATE",
            "name": "expirationDate",
          },
          "numberOfInsureds": Object {
            "cardinality": "MULTIPLE",
            "fieldPath": "numberOfInsureds",
            "fieldType": "INTEGER",
            "name": "numberOfInsureds",
          },
          "oosProcessingStage": Object {
            "cardinality": "SINGLE",
            "fieldPath": "policyDetail.oosProcessingStage",
            "fieldType": "STRING",
            "name": "oosProcessingStage",
          },
          "parties": Object {
            "cardinality": "MULTIPLE",
            "fieldPath": "parties",
            "fieldType": "Party",
            "name": "parties",
          },
          "policies": Object {
            "cardinality": "MULTIPLE",
            "fieldPath": "policies",
            "fieldType": "STRING",
            "name": "policies",
          },
          "policyCurrency": Object {
            "cardinality": "SINGLE",
            "fieldPath": "policyCurrency",
            "fieldType": "STRING",
            "name": "policyCurrency",
          },
          "policyNumber": Object {
            "cardinality": "SINGLE",
            "fieldPath": "policyNumber",
            "fieldType": "STRING",
            "name": "policyNumber",
          },
          "policyValue": Object {
            "cardinality": "SINGLE",
            "fieldPath": "policyValue",
            "fieldType": "MONEY",
            "name": "policyValue",
          },
          "refToCustomer": Object {
            "cardinality": "SINGLE",
            "fieldPath": "refToCustomer",
            "fieldType": "UNKNOWN",
            "name": "refToCustomer",
          },
          "referer": Object {
            "cardinality": "SINGLE",
            "fieldPath": "referer",
            "fieldType": "Referer",
            "name": "referer",
          },
          "riskItems": Object {
            "cardinality": "MULTIPLE",
            "fieldPath": "riskItems",
            "fieldType": "Vehicle",
            "name": "riskItems",
          },
          "state": Object {
            "cardinality": "SINGLE",
            "fieldPath": "state",
            "fieldType": "STRING",
            "name": "state",
          },
          "systemDate": Object {
            "cardinality": "SINGLE",
            "fieldPath": "systemDate",
            "fieldType": "SysDate",
            "name": "systemDate",
          },
          "termCd": Object {
            "cardinality": "SINGLE",
            "fieldPath": "termDetails.termCd",
            "fieldType": "STRING",
            "name": "termCd",
          },
          "termNo": Object {
            "cardinality": "SINGLE",
            "fieldPath": "termDetails.termNo",
            "fieldType": "INTEGER",
            "name": "termNo",
          },
          "totalLimit": Object {
            "cardinality": "SINGLE",
            "fieldPath": "transactionDetails.totalLimit",
            "fieldType": "DECIMAL",
            "name": "totalLimit",
          },
          "totalPremium": Object {
            "cardinality": "SINGLE",
            "fieldPath": "transactionDetails.totalPremium",
            "fieldType": "DECIMAL",
            "name": "totalPremium",
          },
          "txCreateDate": Object {
            "cardinality": "SINGLE",
            "fieldPath": "transactionDetails.txCreateDate",
            "fieldType": "DATE",
            "name": "txCreateDate",
          },
          "txEffectiveDate": Object {
            "cardinality": "SINGLE",
            "fieldPath": "transactionDetails.txEffectiveDate",
            "fieldType": "DATETIME",
            "name": "txEffectiveDate",
          },
          "txReason": Object {
            "cardinality": "SINGLE",
            "fieldPath": "transactionDetails.txReason",
            "fieldType": "STRING",
            "name": "txReason",
          },
          "txType": Object {
            "cardinality": "SINGLE",
            "fieldPath": "transactionDetails.txType",
            "fieldType": "STRING",
            "name": "txType",
          },
          "updatedBy": Object {
            "cardinality": "SINGLE",
            "fieldPath": "accessTrackInfo.updatedBy",
            "fieldType": "STRING",
            "name": "updatedBy",
          },
          "updatedOn": Object {
            "cardinality": "SINGLE",
            "fieldPath": "accessTrackInfo.updatedOn",
            "fieldType": "DATE",
            "name": "updatedOn",
          },
          "versionDescription": Object {
            "cardinality": "SINGLE",
            "fieldPath": "policyDetail.versionDescription",
            "fieldType": "STRING",
            "name": "versionDescription",
          },
        },
        "inheritedContexts": Array [],
        "name": "Policy",
        "system": false,
      },
      "contextId": "0",
      "contextName": "Policy",
      "contextPath": "path.to.0",
      "dataContextReferences": Object {
        "Policy": Object {
          "cardinality": "SINGLE",
          "dataContexts": Array [
            [Circular],
          ],
          "name": "Policy",
        },
      },
      "dataObject": Object {
        "accessTrackInfo": Object {
          "cd": "AccessTrackInfo",
          "id": "9",
        },
        "billingInfo": Object {
          "cd": "BillingInfo",
          "creditCardInfo": Object {
            "cd": "CreditCardInfo",
            "id": "2",
          },
          "id": "1",
        },
        "cd": "Policy",
        "id": "0",
        "insured": Object {
          "addressInfo": Object {
            "cd": "AddressInfo",
            "id": "iai1",
          },
          "cd": "Insured",
          "id": "insured-1-id",
        },
        "parties": Array [
          Object {
            "cd": "Party",
            "id": "3",
            "personInfo": Object {
              "cd": "PersonInfo",
              "id": "4",
            },
            "roles": Array [
              Object {
                "cd": "PartyRole",
                "id": "5",
              },
            ],
          },
        ],
        "policyDetail": Object {
          "cd": "PolicyDetail",
          "id": "11",
        },
        "riskItems": Array [
          Object {
            "addressInfo": Object {
              "cd": "AddressInfo",
              "id": "7",
            },
            "cd": "Vehicle",
            "id": "6",
          },
        ],
        "termDetails": Object {},
        "transactionDetails": Object {
          "cd": "TransactionDetails",
          "id": "8",
        },
      },
      "description": "Policy:path.to.0:0",
      "id": "Policy:0",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "objectReferences": Object {
        "Policy": Object {
          "accessTrackInfo": Object {
            "cd": "AccessTrackInfo",
            "id": "9",
          },
          "billingInfo": Object {
            "cd": "BillingInfo",
            "creditCardInfo": Object {
              "cd": "CreditCardInfo",
              "id": "2",
            },
            "id": "1",
          },
          "cd": "Policy",
          "id": "0",
          "insured": Object {
            "addressInfo": Object {
              "cd": "AddressInfo",
              "id": "iai1",
            },
            "cd": "Insured",
            "id": "insured-1-id",
          },
          "parties": Array [
            Object {
              "cd": "Party",
              "id": "3",
              "personInfo": Object {
                "cd": "PersonInfo",
                "id": "4",
              },
              "roles": Array [
                Object {
                  "cd": "PartyRole",
                  "id": "5",
                },
              ],
            },
          ],
          "policyDetail": Object {
            "cd": "PolicyDetail",
            "id": "11",
          },
          "riskItems": Array [
            Object {
              "addressInfo": Object {
                "cd": "AddressInfo",
                "id": "7",
              },
              "cd": "Vehicle",
              "id": "6",
            },
          ],
          "termDetails": Object {},
          "transactionDetails": Object {
            "cd": "TransactionDetails",
            "id": "8",
          },
        },
      },
      "parent": undefined,
    },
  },
}
`;
